In [1]:
import pandas as pd
from pathlib import Path

In [2]:
data_path = Path("../data")

In [3]:
df_withdrawals = pd.read_parquet(data_path / "zrive_advertiser_withdrawals.parquet")
df_advertiser = pd.read_parquet(data_path / "zrive_dim_advertiser.parquet")
df_monthly = pd.read_parquet(data_path / "zrive_fct_monthly_snapshot_advertiser.parquet")

In [4]:
df_advertiser[df_advertiser['advertiser_zrive_id'] == 8]

advertiser_zrive_id  province_id          updated_at  \
4960                    8           30 2024-02-08 11:21:59   

     advertiser_province  advertiser_group_id min_start_contrato_date  \
4960              Málaga                  NaN              2023-05-23   

     max_start_contrato_nuevo_date contrato_churn_date  
4960                    2023-05-23          2024-01-31

In [5]:
df_withdrawals[df_withdrawals['advertiser_zrive_id'] == 8]

withdrawal_id  advertiser_zrive_id withdrawal_status withdrawal_type  \
12                68                    8           Cerrada           TOTAL   
49               300                    8           Cerrada           TOTAL   
300             1625                    8           Cerrada           TOTAL   
824             4215                    8           Cerrada           TOTAL   
1961           17137                    8           Cerrada           TOTAL   
2002           17418                    8           Cerrada         PARCIAL   
2078           17733                    8           Cerrada           TOTAL   
2129           17974                    8           Cerrada         PARCIAL   
2216           18296                    8           Cerrada           TOTAL   
2315           18718                    8           Cerrada           TOTAL   
2504           19568                    8           Cerrada           TOTAL   
2670           20342                    8           Cerrada           TOTAL   
3056           22009                    8           Cerrada           TOTAL   
3862           24667                    8           Cerrada           TOTAL   
4649           26831                    8           Cerrada           TOTAL   
4781           27146                    8           Cerrada           TOTAL   
4984           27629                    8           Cerrada           TOTAL   
6541           31626                    8           Cerrada           TOTAL   
6856           32485                    8           Cerrada           TOTAL   
7826           34772                    8           Cerrada           TOTAL   
8458           36392                    8           Cerrada           TOTAL   
9530           38530                    8           Cerrada           TOTAL   
9876           39145                    8           Cerrada           TOTAL   
9946           39256                    8           Cerrada           TOTAL   
15060          46441                    8           Cerrada           TOTAL   
15546          47040                    8           Cerrada           TOTAL   

      withdrawal_creation_date                      withdrawal_reason  
12         2012-06-28 09:39:59                     RAZONES ECONOMICAS  
49         2012-09-12 07:24:43                     RAZONES ECONOMICAS  
300        2013-07-09 09:32:41                                  OTROS  
824        2014-09-19 09:12:10                     RAZONES ECONOMICAS  
1961       2016-10-19 14:22:57                              MOROSIDAD  
2002       2016-11-15 08:58:44                    FALTA DE USO/TIEMPO  
2078       2016-12-20 14:21:21                              MOROSIDAD  
2129       2017-01-18 17:13:33                     RAZONES ECONOMICAS  
2216       2017-02-16 10:59:03                              MOROSIDAD  
2315       2017-03-21 11:45:34                     RAZONES ECONOMICAS  
2504       2017-06-21 14:15:13                              MOROSIDAD  
2670       2017-10-10 08:51:11                     RAZONES ECONOMICAS  
3056       2018-05-08 14:40:47                             RESULTADOS  
3862       2019-03-28 11:03:52                             RESULTADOS  
4649       2019-11-11 11:36:12                 Cambio a Bundle Online  
4781       2019-11-28 08:21:25                     RAZONES ECONOMICAS  
4984       2020-01-14 15:34:39                     RAZONES ECONOMICAS  
6541       2020-09-05 10:58:58                     RAZONES ECONOMICAS  
6856       2020-10-27 10:43:52                     RAZONES ECONOMICAS  
7826       2021-04-09 07:32:30                     RAZONES ECONOMICAS  
8458       2021-10-01 08:50:30                 Cambio a Bundle Online  
9530       2022-04-18 07:34:08                             RESULTADOS  
9876       2022-06-03 09:03:54  Cambio de Contrato/propuesta/producto  
9946       2022-06-15 08:00:13                             RESULTADOS  
15060      2023-10-20 11:35:22                              MOROSIDAD  
1

In [11]:
df_monthly[df_monthly['advertiser_zrive_id'] == 8]

advertiser_zrive_id  period_int  monthly_published_ads  \
13752                    8      202305                      9   
17141                    8      202306                     50   
20502                    8      202307                     50   
23941                    8      202308                     60   
27165                    8      202309                     54   
30365                    8      202310                     50   
33509                    8      202311                     50   
36725                    8      202312                     50   
39931                    8      202401                     47   

       monthly_unique_published_ads  monthly_contracted_ads  monthly_leads  \
13752                             9                      50              2   
17141                            50                      50             21   
20502                            50                      50             29   
23941                            60                      50             26   
27165                            54                      50             34   
30365                            50                      50             29   
33509                            50                      50             12   
36725                            50                      50             27   
39931                            47                      50             17   

       monthly_visits  monthly_oro_ads  monthly_plata_ads  \
13752          5542.5                1                  1   
17141         41137.5                1                  1   
20502         39435.0                1                  1   
23941         31350.0                1                  1   
27165         41850.0                1                  1   
30365         39435.0                1                  1   
33509         25327.5                1                  1   
36725         37965.0                1                  1   
39931         82582.5                1                  1   

       monthly_destacados_ads  ...  monthly_total_phone_views  \
13752                       1  ...                          1   
17141                       1  ...                         19   
20502                       1  ...                         23   
23941                       1  ...                         14   
27165                       1  ...                         25   
30365                       1  ...                         31   
33509                       1  ...                         12   
36725                       1  ...                         22   
39931                       1  ...                         20   

       monthly_total_calls  monthly_total_emails  monthly_total_invoice  \
13752                    0                     1                    4.3   
17141                    9                     9                  107.1   
20502                    9                    15                  107.1   
23941                   10                    10                  107.1   
27165                   15                     9                  213.9   
30365                    9                    10                  213.9   
33509                    6                     6                   43.3   
36725                   16                     7                   43.3   
39931                    7                     7                  128.6   

       monthly_unique_calls  monthly_unique_emails  monthly_unique_leads  \
13752                     0                      2                     2   
17141                     7                     11                    18   
20502                     8                     11                    19   
23941                     5                     13                    18   
27165                     6                     17                    23   
30365                     8                     20                    28   
33509                     5   

In [157]:
df_withdrawals[df_withdrawals['advertiser_zrive_id'] == 21]

withdrawal_id  advertiser_zrive_id withdrawal_status withdrawal_type  \
7464           33933                   21           Cerrada           TOTAL   
10968          41025                   21           Cerrada           TOTAL   
12393          42883                   21           Cerrada           TOTAL   
15290          46732                   21           Cerrada           TOTAL   
15630          47128                   21           Cerrada           TOTAL   
15807          47353                   21           Cerrada           TOTAL   
16668          48409                   21           Cerrada           TOTAL   
16874          48688                   21           Cerrada           TOTAL   
18361          50419                   21           Cerrada           TOTAL   
20092          52362                   21           Cerrada           TOTAL   

      withdrawal_creation_date                      withdrawal_reason  
7464       2021-01-28 11:47:43                                  OTROS  
10968      2022-11-14 15:29:19  Cambio de Contrato/propuesta/producto  
12393      2023-03-06 12:14:02  Cambio de Contrato/propuesta/producto  
15290      2023-11-06 16:58:13      Reestructuración cuentas de grupo  
15630      2023-12-05 11:43:09              RATIO RESULTADO-INVERSION  
15807      2023-12-20 15:25:02  Cambio de Contrato/propuesta/producto  
16668      2024-02-29 17:11:33                             RESULTADOS  
16874      2024-03-15 08:37:10           Upselling-cambio de contrato  
18361      2024-06-28 11:04:34  Cambio de Contrato/propuesta/producto  
20092      2024-11-11 05:54:18  Cambio de Contrato/propuesta/producto

In [179]:
df_withdrawals[df_withdrawals['withdrawal_id'] == 48688]

withdrawal_id  advertiser_zrive_id withdrawal_status withdrawal_type  \
16874          48688                   21           Cerrada           TOTAL   

      withdrawal_creation_date             withdrawal_reason  
16874      2024-03-15 08:37:10  Upselling-cambio de contrato

In [159]:
df_advertiser[df_advertiser['advertiser_zrive_id'] == 21]

advertiser_zrive_id  province_id          updated_at  \
3001                   21           46 2024-06-05 16:01:08   

     advertiser_province  advertiser_group_id min_start_contrato_date  \
3001            Valencia                  NaN              2022-11-14   

     max_start_contrato_nuevo_date contrato_churn_date  
3001                          None                None

In [160]:
df_advertiser['advertiser_zrive_id'].value_counts()

advertiser_zrive_id
6732    1
3229    1
2921    1
2805    1
451     1
       ..
344     1
3168    1
2573    1
6242    1
3586    1
Name: count, Length: 6834, dtype: int64

In [161]:
df_withdrawals.head()

withdrawal_id  advertiser_zrive_id withdrawal_status withdrawal_type  \
0              0                  257           Cerrada           TOTAL   
1              1                  219           Cerrada           TOTAL   
2              7                  487           Cerrada           TOTAL   
3             12                  476           Cerrada           TOTAL   
4             16                  452           Cerrada           TOTAL   

  withdrawal_creation_date    withdrawal_reason  
0      2012-06-19 07:12:34           RESULTADOS  
1      2012-06-19 07:16:34           RESULTADOS  
2      2012-06-20 07:10:16           RESULTADOS  
3      2012-06-20 11:59:36           RESULTADOS  
4      2012-06-20 15:41:39  FALTA DE USO/TIEMPO

In [6]:
# Withdrawals 
def add_churn(df: pd.DataFrame):
    CHURN_REASONS_EXCLUDED = [
        'Upselling-cambio de contrato',
        'Cambio a Bundle Online',
        'Cambio de Contrato/propuesta/producto'
    ]
    df["churn"] = (
        (df["withdrawal_type"] == "TOTAL") &
        (df["withdrawal_status"] != "Denegada") &
        (~df["withdrawal_reason"].isin(CHURN_REASONS_EXCLUDED))
    ).astype(int)
    return df

def convert_datetime_to_month_period(df, datetime_col, new_col, drop_original=True):
    df[new_col] = pd.to_datetime(df[datetime_col]).dt.to_period('M')
    if drop_original:
        df = df.drop(columns=[datetime_col])
    return df

def add_predict_month(df: pd.DataFrame, predict_col = "predict_month", withdrawal_col="withdrawal_month", n: int = 1):
    df[predict_col] = df[withdrawal_col] - n
    return df


In [7]:
WITHDRAWAL_COLS_TO_DROP = ["withdrawal_id", "withdrawal_status", "withdrawal_type", "withdrawal_reason", "withdrawal_month"]

df_withdrawals = convert_datetime_to_month_period(
    df_withdrawals, 
    datetime_col='withdrawal_creation_date',
    new_col='withdrawal_month',
    drop_original=True
)
df_withdrawals = add_predict_month(df_withdrawals, n=1)
df_withdrawals = add_churn(df_withdrawals)
df_withdrawals.drop(columns=WITHDRAWAL_COLS_TO_DROP, inplace=True)

In [8]:
df_withdrawals

advertiser_zrive_id predict_month  churn
0                      257       2012-05      1
1                      219       2012-05      1
2                      487       2012-05      1
3                      476       2012-05      1
4                      452       2012-05      1
...                    ...           ...    ...
20674                 5441       2024-11      1
20675                 5439       2024-11      1
20676                  154       2024-11      1
20677                 1352       2024-11      1
20678                 4153       2024-11      1

[20679 rows x 3 columns]

In [9]:
df_withdrawals[df_withdrawals['advertiser_zrive_id'] == 21]

advertiser_zrive_id predict_month  churn
7464                    21       2020-12      1
10968                   21       2022-10      0
12393                   21       2023-02      0
15290                   21       2023-10      1
15630                   21       2023-11      1
15807                   21       2023-11      0
16668                   21       2024-01      1
16874                   21       2024-02      0
18361                   21       2024-05      0
20092                   21       2024-10      0

In [10]:
df_withdrawals[df_withdrawals['advertiser_zrive_id'] == 8]

advertiser_zrive_id predict_month  churn
12                       8       2012-05      1
49                       8       2012-08      1
300                      8       2013-06      1
824                      8       2014-08      1
1961                     8       2016-09      1
2002                     8       2016-10      0
2078                     8       2016-11      1
2129                     8       2016-12      0
2216                     8       2017-01      1
2315                     8       2017-02      1
2504                     8       2017-05      1
2670                     8       2017-09      1
3056                     8       2018-04      1
3862                     8       2019-02      1
4649                     8       2019-10      0
4781                     8       2019-10      1
4984                     8       2019-12      1
6541                     8       2020-08      1
6856                     8       2020-09      1
7826                     8       2021-03      1
8458                     8       2021-09      0
9530                     8       2022-03      1
9876                     8       2022-05      0
9946                     8       2022-05      1
15060                    8       2023-09      1
15546                    8       2023-10      1

### df_monthly --> como definir churn

In [166]:
df_monthly.columns

Index(['advertiser_zrive_id', 'period_int', 'monthly_published_ads',
       'monthly_unique_published_ads', 'monthly_contracted_ads',
       'monthly_leads', 'monthly_visits', 'monthly_oro_ads',
       'monthly_plata_ads', 'monthly_destacados_ads', 'monthly_pepitas_ads',
       'monthly_shows', 'monthly_total_phone_views', 'monthly_total_calls',
       'monthly_total_emails', 'monthly_total_invoice', 'monthly_unique_calls',
       'monthly_unique_emails', 'monthly_unique_leads', 'monthly_avg_ad_price',
       'monthly_distinct_ads', 'has_active_contract'],
      dtype='object')

In [167]:
df_monthly[df_monthly['advertiser_zrive_id'] == 3]

advertiser_zrive_id  period_int  monthly_published_ads  \
2                        3      202301                      0   
53018                    3      202405                     17   
56458                    3      202406                     17   
59987                    3      202407                     17   

       monthly_unique_published_ads  monthly_contracted_ads  monthly_leads  \
2                                 0                       0              0   
53018                            17                      35              6   
56458                            17                      35              0   
59987                            17                      35              0   

       monthly_visits  monthly_oro_ads  monthly_plata_ads  \
2                 0.0                0                  0   
53018          2167.5                0                  0   
56458          1882.5                0                  0   
59987           315.0                0                  0   

       monthly_destacados_ads  ...  monthly_total_phone_views  \
2                           0  ...                          0   
53018                       0  ...                          2   
56458                       0  ...                          0   
59987                       0  ...                          0   

       monthly_total_calls  monthly_total_emails  monthly_total_invoice  \
2                        0                     0                    0.0   
53018                    1                     3                    0.0   
56458                    0                     0                    0.0   
59987                    0                     0                    0.0   

       monthly_unique_calls  monthly_unique_emails  monthly_unique_leads  \
2                         0                      0                     0   
53018                     1                      2                     3   
56458                     0                      0                     0   
59987                     0                      0                     0   

       monthly_avg_ad_price  monthly_distinct_ads  has_active_contract  
2                       NaN                   NaN                False  
53018              42864.59                  17.0                 True  
56458              42864.59                  17.0                 True  
59987              42864.59                  17.0                 True  

[4 rows x 22 columns]

In [168]:
def convert_period_int_to_month_period(df, period_col='period_int', new_col='month_period'):
    """Convierte el formato YYYYMM a un período mensual de pandas"""
    df[new_col] = pd.to_datetime(df[period_col].astype(str) + '01', format='%Y%m%d').dt.to_period('M')
    return df

In [169]:
df_monthly = convert_period_int_to_month_period(df_monthly)

In [170]:
df_monthly.head()

advertiser_zrive_id  period_int  monthly_published_ads  \
0                    1      202301                     47   
1                    2      202301                     31   
2                    3      202301                      0   
3                    4      202301                     79   
4                    6      202301                     20   

   monthly_unique_published_ads  monthly_contracted_ads  monthly_leads  \
0                            47                      75             18   
1                            31                     150              4   
2                             0                       0              0   
3                            79                      85             14   
4                            20                      20             16   

   monthly_visits  monthly_oro_ads  monthly_plata_ads  monthly_destacados_ads  \
0         40890.0                6                  6                       6   
1         17970.0               10                 10                       4   
2             0.0                0                  0                       0   
3         27157.5                3                  3                       1   
4         79492.5                0                  0                       1   

   ...  monthly_total_calls  monthly_total_emails  monthly_total_invoice  \
0  ...                   15                     0                  440.8   
1  ...                    2                     2                   75.4   
2  ...                    0                     0                    0.0   
3  ...                    8                     2                  299.6   
4  ...                    4                     3                   86.5   

   monthly_unique_calls  monthly_unique_emails  monthly_unique_leads  \
0                    12                      3                    15   
1                     2                      2                     4   
2                     0                      0                     0   
3                     6                      6                    12   
4                     4                     11                    15   

   monthly_avg_ad_price  monthly_distinct_ads  has_active_contract  \
0                   NaN                   NaN                 True   
1                   NaN                   NaN                 True   
2                   NaN                   NaN                False   
3                   NaN                   NaN                 True   
4                   NaN                   NaN                 True   

   month_period  
0       2023-01  
1       2023-01  
2       2023-01  
3       2023-01  
4       2023-01  

[5 rows x 23 columns]

In [171]:
def add_churn_target(df_monthly, df_withdrawals):
    """
    Añade el target de churn a df_monthly, usando lo calculado en df_withdrawals

    Args:
        df_monthly: DataFrame con datos mensuales de anunciantes
        df_withdrawals: DataFrame procesado con información de churn
        
    Returns:
        DataFrame con columna 'churn' añadida
    """
    # Realizar la unión basada en advertiser_id y el período mensual
    df_target = df_monthly.merge(
        df_withdrawals[['advertiser_zrive_id', 'predict_month', 'churn']],
        left_on=['advertiser_zrive_id', 'month_period'],
        right_on=['advertiser_zrive_id', 'predict_month'],
        how='left'
    )

    # Rellenar los valores NaN en churn con 0 (no hubo churn)
    df_target['churn'] = df_target['churn'].fillna(0)

    # Eliminar la columna redundante predict_month
    if 'predict_month' in df_target.columns:
        df_target = df_target.drop(columns=['predict_month'])

    return df_target

In [172]:
df_target = add_churn_target(df_monthly, df_withdrawals)

In [173]:
df_target[df_target['advertiser_zrive_id'] == 8]

advertiser_zrive_id  period_int  monthly_published_ads  \
13821                    8      202305                      9   
17224                    8      202306                     50   
20596                    8      202307                     50   
24042                    8      202308                     60   
27285                    8      202309                     54   
30499                    8      202310                     50   
33660                    8      202311                     50   
36889                    8      202312                     50   
40121                    8      202401                     47   

       monthly_unique_published_ads  monthly_contracted_ads  monthly_leads  \
13821                             9                      50              2   
17224                            50                      50             21   
20596                            50                      50             29   
24042                            60                      50             26   
27285                            54                      50             34   
30499                            50                      50             29   
33660                            50                      50             12   
36889                            50                      50             27   
40121                            47                      50             17   

       monthly_visits  monthly_oro_ads  monthly_plata_ads  \
13821          5542.5                1                  1   
17224         41137.5                1                  1   
20596         39435.0                1                  1   
24042         31350.0                1                  1   
27285         41850.0                1                  1   
30499         39435.0                1                  1   
33660         25327.5                1                  1   
36889         37965.0                1                  1   
40121         82582.5                1                  1   

       monthly_destacados_ads  ...  monthly_total_emails  \
13821                       1  ...                     1   
17224                       1  ...                     9   
20596                       1  ...                    15   
24042                       1  ...                    10   
27285                       1  ...                     9   
30499                       1  ...                    10   
33660                       1  ...                     6   
36889                       1  ...                     7   
40121                       1  ...                     7   

       monthly_total_invoice  monthly_unique_calls  monthly_unique_emails  \
13821                    4.3                     0                      2   
17224                  107.1                     7                     11   
20596                  107.1                     8                     11   
24042                  107.1                     5                     13   
27285                  213.9                     6                     17   
30499                  213.9                     8                     20   
33660                   43.3                     5                      6   
36889                   43.3                    11                      9   
40121                  128.6                     6                      8   

       monthly_unique_leads  monthly_avg_ad_price  monthly_distinct_ads  \
13821                     2                   NaN                   NaN   
17224                    18                   NaN                   NaN   
20596                    19                   NaN                   NaN   
24042                    18                   NaN                   NaN   
27285                    23                   NaN                   NaN   
30499                    28              16329.90                  40.0   
33660                    11              12519.28                  

In [174]:
def reduce_columns(df):
    """
    Selecciona solo ciertas columnas para el análisis inicial
    """
    essential_columns = [
        'advertiser_zrive_id',
        'period_int',
        'month_period',
        'monthly_published_ads',
        'monthly_contracted_ads',
        'has_active_contract',
        'churn'
    ]

    # Asegurarse de que todas las columnas existen
    existing_columns = [col for col in essential_columns if col in df.columns]

    return df[existing_columns]

In [175]:
df_target = reduce_columns(df_target)

In [176]:
df_target[df_target['advertiser_zrive_id'] == 8]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
13821                    8      202305      2023-05                      9   
17224                    8      202306      2023-06                     50   
20596                    8      202307      2023-07                     50   
24042                    8      202308      2023-08                     60   
27285                    8      202309      2023-09                     54   
30499                    8      202310      2023-10                     50   
33660                    8      202311      2023-11                     50   
36889                    8      202312      2023-12                     50   
40121                    8      202401      2024-01                     47   

       monthly_contracted_ads  has_active_contract  churn  
13821                      50                 True    0.0  
17224                      50                 True    0.0  
20596                      50                 True    0.0  
24042                      50                 True    0.0  
27285                      50                 True    1.0  
30499                      50                 True    1.0  
33660                      50                 True    0.0  
36889                      50                 True    0.0  
40121                      50                 True    0.0

Creo que es un buen punto de partida. 

- Creo que deberiamos eliminar las filas que hay despues de la fila con churn=1

- Hay varios casos que veo que pueden darnos problemas

    - Caso 1: Las filas que tienen has_active_contract en False que significan? Yo entiendo que no tienen contrato y que por tanto no pueden tener interacciones ese mes pero si que tienen, que hacemos con estas filas las dejamos o las quitamos?


    - Caso 2: Usuarios que dejan de tener actividad pero no tienen churn confirmada en df_withdrawals entre 2023 y 2024

    - Caso 3: Usuarios con multiples churn

### Caso 1: has_active_contract en False 

In [40]:
df_target[df_target['has_active_contract'] == False].head(15)

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
2                       3      202301      2023-01                      0   
29                     41      202301      2023-01                     52   
55                     76      202301      2023-01                     20   
503                   693      202301      2023-01                     22   
574                   787      202301      2023-01                     16   
585                   803      202301      2023-01                     13   
661                   917      202301      2023-01                     27   
775                  1078      202301      2023-01                     75   
897                  1246      202301      2023-01                     15   
952                  1333      202301      2023-01                     34   
986                  1373      202301      2023-01                      0   
992                  1380      202301      2023-01                     16   
1279                 1763      202301      2023-01                     13   
1586                 2215      202301      2023-01                     35   
1685                 2353      202301      2023-01                     45   

      monthly_contracted_ads  has_active_contract  churn  
2                          0                False    0.0  
29                       100                False    0.0  
55                        75                False    0.0  
503                       35                False    0.0  
574                      200                False    0.0  
585                       20                False    0.0  
661                       35                False    0.0  
775                       75                False    0.0  
897                       20                False    0.0  
952                       35                False    0.0  
986                       15                False    0.0  
992                       20                False    0.0  
1279                      75                False    0.0  
1586                      35                False    0.0  
1685                     400                False    0.0

In [41]:
#Caso 1: Ejemplo

df_target[df_target['advertiser_zrive_id'] == 2353]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
1685                  2353      202301      2023-01                     45   
5137                  2353      202302      2023-02                    400   
8527                  2353      202303      2023-03                    400   
11932                 2353      202304      2023-04                    400   
15360                 2353      202305      2023-05                    426   
18742                 2353      202306      2023-06                    290   
22129                 2353      202307      2023-07                    300   
25501                 2353      202308      2023-08                    300   

       monthly_contracted_ads  has_active_contract  churn  
1685                      400                False    0.0  
5137                      400                False    0.0  
8527                      400                 True    0.0  
11932                     400                 True    1.0  
15360                     400                 True    0.0  
18742                     400                 True    0.0  
22129                     400                 True    0.0  
25501                     400                False    0.0

In [42]:
#Caso 1: Ejemplo

df_target[df_target['advertiser_zrive_id'] == 76]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
55                      76      202301      2023-01                     20   
3605                    76      202302      2023-02                     29   
7002                    76      202303      2023-03                     25   
10433                   76      202304      2023-04                     14   
13866                   76      202305      2023-05                     29   

       monthly_contracted_ads  has_active_contract  churn  
55                         75                False    0.0  
3605                       75                False    0.0  
7002                       75                False    0.0  
10433                      75                 True    1.0  
13866                     150                 True    0.0

In [43]:
df_target[df_target['advertiser_zrive_id'] == 787]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
574                    787      202301      2023-01                     16   
30992                  787      202310      2023-10                     20   
34162                  787      202311      2023-11                     20   
37392                  787      202312      2023-12                     20   
40618                  787      202401      2024-01                     35   
43836                  787      202402      2024-02                     35   
47092                  787      202403      2024-03                     35   
50422                  787      202404      2024-04                     35   
53802                  787      202405      2024-05                     20   
57264                  787      202406      2024-06                     20   
60796                  787      202407      2024-07                     20   

       monthly_contracted_ads  has_active_contract  churn  
574                       200                False    0.0  
30992                      20                 True    0.0  
34162                      20                 True    0.0  
37392                      20                 True    0.0  
40618                      35                 True    0.0  
43836                      35                 True    0.0  
47092                      35                 True    0.0  
50422                      35                 True    1.0  
53802                      20                 True    0.0  
57264                      20                 True    0.0  
60796                      20                 True    0.0

In [44]:
df_target['has_active_contract'].value_counts()

has_active_contract
True     78452
False     2112
Name: count, dtype: int64

Puede que sea un periodo de prueba gratuito, ya que siempre es en los primeros meses. Lo mas probable es que en las filas que tienen esto nunca haga churn y es lo que aprendera el modelo. Creo que no hay problema con los datos de has_Active_contract = False y lo podemos dejar tal cual esta

### Caso 2: Usuarios que dejan de tener actividad pero no tienen churn confirmada en df_withdrawals entre 2023 y 2024

In [45]:
#Caso 2: Ejemplo

df_target[df_target['advertiser_zrive_id'] == 17]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
11                     17      202301      2023-01                     15   
3563                   17      202302      2023-02                     14   
6958                   17      202303      2023-03                      0   

      monthly_contracted_ads  has_active_contract  churn  
11                        35                 True    0.0  
3563                      35                 True    0.0  
6958                      35                 True    0.0

In [46]:
df_withdrawals[df_withdrawals['advertiser_zrive_id'] == 17]

advertiser_zrive_id predict_month  churn
11367                   17       2022-11      1

En este caso intuyo que el churn es en 2022-11 (fila que tendria churn=1, pero que como es de 2022 no la tenemos) y el resto de filas son la actividad del contrato que ya la  tenia pagado, pero que eliminamos con el enfoque de "eliminar las filas que hay despues de la fila con churn=1"

In [47]:
df_target[df_target['advertiser_zrive_id'] == 46]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
34                   46      202301      2023-01                    100   

    monthly_contracted_ads  has_active_contract  churn  
34                     100                 True    0.0

In [48]:
df_withdrawals[df_withdrawals['advertiser_zrive_id'] == 46]

advertiser_zrive_id predict_month  churn
8009                    46       2021-04      1
11865                   46       2022-12      1

In [49]:
df_target[df_target['advertiser_zrive_id'] == 176]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
136                   176      202301      2023-01                     11   
3682                  176      202302      2023-02                     11   

      monthly_contracted_ads  has_active_contract  churn  
136                       20                 True    0.0  
3682                      20                 True    0.0

In [50]:
df_withdrawals[df_withdrawals['advertiser_zrive_id'] == 176]

advertiser_zrive_id predict_month  churn
3360                   176       2018-09      1
6448                   176       2020-06      1
9489                   176       2022-03      1
10387                  176       2022-08      1
11048                  176       2022-10      1

In [51]:
df_target[df_target['advertiser_zrive_id'] == 18]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
12                     18      202301      2023-01                      1   
3564                   18      202302      2023-02                      1   
6959                   18      202303      2023-03                      1   

      monthly_contracted_ads  has_active_contract  churn  
12                        50                 True    0.0  
3564                      50                 True    0.0  
6959                      50                False    0.0

In [52]:
df_withdrawals[df_withdrawals['advertiser_zrive_id'] == 18]

advertiser_zrive_id predict_month  churn
11471                   18       2022-11      1

En estos casos ocurre lo mismo

In [53]:
df_target[df_target['advertiser_zrive_id'] == 19]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
13827                   19      202305      2023-05                     10   
17230                   19      202306      2023-06                     15   
20602                   19      202307      2023-07                     15   

       monthly_contracted_ads  has_active_contract  churn  
13827                      50                 True    0.0  
17230                      50                 True    0.0  
20602                      50                 True    0.0

In [54]:
df_withdrawals[df_withdrawals['advertiser_zrive_id'] == 19]

advertiser_zrive_id predict_month  churn
1354                   19       2015-09      1
3068                   19       2018-04      1

Pero hay otros casos que no son asi, realmente son churn, pero no queda registrada, entiendo que es a lo que nos referimos en el doc de google de " Deja de haber actividad a partir de cierto momento". Aqui el problema es que sabemos cuando se le acabo el contrato pero no cuando fue el churn, el enfoque que hagamos va a ser una aproximacion.  Ver a cuantos usuarios afecta este problema

In [55]:
user_churn_status = df_target.groupby('advertiser_zrive_id')['churn'].max().reset_index()
# 1. Identificar usuarios sin churn
users_without_churn_ids = user_churn_status[user_churn_status['churn'] == 0]['advertiser_zrive_id'].tolist()

# 2. Analizar el último período registrado para cada usuario sin churn
last_period_by_user = df_target.groupby('advertiser_zrive_id')['period_int'].max().reset_index()
last_period_no_churn = last_period_by_user[last_period_by_user['advertiser_zrive_id'].isin(users_without_churn_ids)]

# 3. Distribución resumida de los últimos períodos para usuarios sin churn
latest_period = df_target['period_int'].max()
users_active_until_end = last_period_no_churn[last_period_no_churn['period_int'] == latest_period].shape[0]
early_ending_users = last_period_no_churn[last_period_no_churn['period_int'] < latest_period].shape[0]

print(f"Usuarios sin churn activos hasta el último período ({latest_period}): {users_active_until_end}")
print(f"Usuarios sin churn que terminaron antes del último período: {early_ending_users}")

# 4. Verificar el estado del contrato en el último registro para algunos usuarios que terminaron antes
early_ending_sample = last_period_no_churn[last_period_no_churn['period_int'] < latest_period]['advertiser_zrive_id'].head(5).tolist()
early_ending_details = []

for user_id in early_ending_sample:
    last_record = df_target[df_target['advertiser_zrive_id'] == user_id].sort_values('period_int').iloc[-1]
    early_ending_details.append({
        'advertiser_zrive_id': user_id,
        'last_period': last_record['period_int'],
        'has_active_contract': last_record['has_active_contract']
    })

#print("\nEjemplos de usuarios sin churn que terminaron antes:")
#print(pd.DataFrame(early_ending_details))

# 3. Distribución de los últimos períodos para usuarios sin churn
print("\n\n")
last_period_distribution = last_period_no_churn['period_int'].value_counts().sort_index()

print("Distribución de últimos períodos para usuarios sin churn:")
print(last_period_distribution)

Usuarios sin churn activos hasta el último período (202412): 2109
Usuarios sin churn que terminaron antes del último período: 563



Distribución de últimos períodos para usuarios sin churn:
period_int
202301     237
202302      85
202303      61
202304      20
202305      20
202306      11
202307      19
202308      10
202309       5
202310       7
202311      10
202312       7
202401      11
202402       5
202403       3
202404       2
202405       8
202406       8
202407       4
202408       7
202409      11
202410       8
202411       4
202412    2109
Name: count, dtype: int64


Tiene sentido que en los meses que mas ocurra esto sea en los de inicio de 2023, ya que seran casos similares al id 17, 46 o 176 que veiamos arriba. Sin embargo hay casos como el del id 19 que no tienen esa logica

Opciones:
- Eliminar estos casos ya que, la mayoria son realmente churns de finales de 2022 y esta es la actividad posterior que eliminamos con el enfoque propuesto.
En el resto de meses ocurre pero en muy pocos casos, creo que no merece la pena definir un nuevo enfoque para incluir tan pocos casos

### Caso 3: Usuarios con multiples churn

In [56]:
# Número usuarios únicos
total_users = df_target['advertiser_zrive_id'].nunique()
print(f"Total de usuarios únicos: {total_users}")

# Usuarios sin churn y con churn
user_churn_status = df_target.groupby('advertiser_zrive_id')['churn'].max().reset_index()
users_without_churn = user_churn_status[user_churn_status['churn'] == 0].shape[0]
users_with_churn = user_churn_status[user_churn_status['churn'] == 1].shape[0]
print(f"Usuarios sin churn: {users_without_churn}")
print(f"Usuarios con al menos un churn: {users_with_churn}")

# Usuarios con uno o múltiples churns
churn_counts = df_target[df_target['churn'] == 1].groupby('advertiser_zrive_id').size()
users_with_one_churn = (churn_counts == 1).sum()
users_with_multiple_churn = (churn_counts > 1).sum()
print(f"Usuarios con exactamente un churn: {users_with_one_churn}")
print(f"Usuarios con múltiples churns: {users_with_multiple_churn}")

Total de usuarios únicos: 6534
Usuarios sin churn: 2672
Usuarios con al menos un churn: 3862
Usuarios con exactamente un churn: 2548
Usuarios con múltiples churns: 1314


In [65]:
df_target[df_target['advertiser_zrive_id'] == 21]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
13                      21      202301      2023-01                     28   
3565                    21      202302      2023-02                     31   
6960                    21      202303      2023-03                     22   
10393                   21      202304      2023-04                     24   
13828                   21      202305      2023-05                     20   
17231                   21      202306      2023-06                     17   
20603                   21      202307      2023-07                     18   
24048                   21      202308      2023-08                     15   
27290                   21      202309      2023-09                     12   
30504                   21      202310      2023-10                     14   
33665                   21      202311      2023-11                      9   
33666                   21      202311      2023-11                      9   
36895                   21      202312      2023-12                      9   
40127                   21      202401      2024-01                     20   
43344                   21      202402      2024-02                     20   
46606                   21      202403      2024-03                     44   
49932                   21      202404      2024-04                     41   
53309                   21      202405      2024-05                     48   
56777                   21      202406      2024-06                     41   
60329                   21      202407      2024-07                     20   
63729                   21      202408      2024-08                     20   
67080                   21      202409      2024-09                     20   
70456                   21      202410      2024-10                     33   
73787                   21      202411      2024-11                     20   
77126                   21      202412      2024-12                     22   

       monthly_contracted_ads  has_active_contract  churn  
13                         35                 True    0.0  
3565                       35                 True    0.0  
6960                       35                 True    0.0  
10393                      35                 True    0.0  
13828                      35                 True    0.0  
17231                      35                 True    0.0  
20603                      35                 True    0.0  
24048                      35                 True    0.0  
27290                      35                 True    0.0  
30504                      35                 True    1.0  
33665                      35                 True    1.0  
33666                      35                 True    0.0  
36895                      20                 True    0.0  
40127                      20                 True    1.0  
43344                      20                 True    0.0  
46606                      50                 True    0.0  
49932                      50                 True    0.0  
53309                      50                 True    0.0  
56777                      50                 True    0.0  
60329                      20                 True    0.0  
63729                      20                 True    0.0  
67080                      20                 True    0.0  
70456                      20                 True    0.0  
73787                     300                 True    0.0  
77126                     300                 True    0.0

Como adaptar el enfoque de "eliminar las filas que hay despues de la fila con churn=1" en estos casos?. 
Quizas se pueda seguir con ese enfoque, pero  añadiendo una columna que sea "numero de churns historicas" o "churn previo" si o no,  y capturar esa info de esta forma (?), habria que tener cuidado y ver que no hacemos data leakage si entrenamos por execution_month

### Caso 1: Solución

Que haya filas con has_Active_contract en False no es un problema, no hay que hacer nada con los datos

### Caso 2: Solución

Creo que lo que mas sentido tiene es eliminar las filas de los 563 usuarios que en su ultimo mes (sin contar 202412), no tienen la etiqueta de Churn=1. Aunque parezcan muchos usuarios, entre enero,febrero y marzo de 2023, suman 383, los cuales son, siguiendo la logica vista, casos de churn a finales de 2022 y lo que vemos es la actividad posterior (no nos interesa en el modelo)

In [58]:
def remove_inactive_users_without_churn(df_target):
    """
    Elimina las filas de usuarios que terminan su actividad antes del último período
    sin tener un churn registrado.
    
    Args:
        df_target: DataFrame con datos mensuales y etiqueta de churn
    
    Returns:
        DataFrame limpio sin estos usuarios
    """
    # Identificar el último período disponible
    latest_period = df_target['period_int'].max()
    
    # Identificar usuarios sin churn
    user_churn_status = df_target.groupby('advertiser_zrive_id')['churn'].max().reset_index()
    users_without_churn = user_churn_status[user_churn_status['churn'] == 0]['advertiser_zrive_id'].tolist()
    
    # Identificar el último período registrado para cada usuario sin churn
    last_period_by_user = df_target.groupby('advertiser_zrive_id')['period_int'].max().reset_index()
    
    # Identificar usuarios sin churn que terminan antes del último período
    early_ending_users = last_period_by_user[
        (last_period_by_user['advertiser_zrive_id'].isin(users_without_churn)) & 
        (last_period_by_user['period_int'] < latest_period)
    ]['advertiser_zrive_id'].tolist()
    
    # Eliminar estos usuarios del dataset
    df_clean = df_target[~df_target['advertiser_zrive_id'].isin(early_ending_users)]
    
    return df_clean

In [59]:
df_target = remove_inactive_users_without_churn(df_target)

In [60]:
user_churn_status = df_target.groupby('advertiser_zrive_id')['churn'].max().reset_index()
# 1. Identificar usuarios sin churn
users_without_churn_ids = user_churn_status[user_churn_status['churn'] == 0]['advertiser_zrive_id'].tolist()

# 2. Analizar el último período registrado para cada usuario sin churn
last_period_by_user = df_target.groupby('advertiser_zrive_id')['period_int'].max().reset_index()
last_period_no_churn = last_period_by_user[last_period_by_user['advertiser_zrive_id'].isin(users_without_churn_ids)]

# 3. Distribución resumida de los últimos períodos para usuarios sin churn
latest_period = df_target['period_int'].max()
users_active_until_end = last_period_no_churn[last_period_no_churn['period_int'] == latest_period].shape[0]
early_ending_users = last_period_no_churn[last_period_no_churn['period_int'] < latest_period].shape[0]

print(f"Usuarios sin churn activos hasta el último período ({latest_period}): {users_active_until_end}")
print(f"Usuarios sin churn que terminaron antes del último período: {early_ending_users}")

# 4. Verificar el estado del contrato en el último registro para algunos usuarios que terminaron antes
early_ending_sample = last_period_no_churn[last_period_no_churn['period_int'] < latest_period]['advertiser_zrive_id'].head(5).tolist()
early_ending_details = []

for user_id in early_ending_sample:
    last_record = df_target[df_target['advertiser_zrive_id'] == user_id].sort_values('period_int').iloc[-1]
    early_ending_details.append({
        'advertiser_zrive_id': user_id,
        'last_period': last_record['period_int'],
        'has_active_contract': last_record['has_active_contract']
    })

#print("\nEjemplos de usuarios sin churn que terminaron antes:")
#print(pd.DataFrame(early_ending_details))

# 3. Distribución de los últimos períodos para usuarios sin churn
print("\n\n")
last_period_distribution = last_period_no_churn['period_int'].value_counts().sort_index()

print("Distribución de últimos períodos para usuarios sin churn:")
print(last_period_distribution)

Usuarios sin churn activos hasta el último período (202412): 2109
Usuarios sin churn que terminaron antes del último período: 0



Distribución de últimos períodos para usuarios sin churn:
period_int
202412    2109
Name: count, dtype: int64


Con este enfoque, los usuarios o tienen churn registrado, o siguen activos hasta el final del período de observación

### Caso 3: Solución

In [61]:
def find_users_with_multiple_churns(df_target):
    """
    Identifica usuarios con múltiples churns en el dataset y muestra ejemplos.

    Args:
        df_target: DataFrame con datos mensuales y etiqueta de churn

    Returns:
        dict: Información sobre usuarios con múltiples churns
    """
    # Contar churns por usuario
    churn_counts = df_target[df_target['churn'] == 1].groupby('advertiser_zrive_id').size()

    # Identificar usuarios con múltiples churns
    multi_churn_users = churn_counts[churn_counts > 1].sort_values(ascending=False)

    # Estadísticas básicas
    total_users = df_target['advertiser_zrive_id'].nunique()
    users_with_churn = churn_counts.shape[0]
    users_with_multiple_churns = multi_churn_users.shape[0]

    print(f"Total de usuarios únicos: {total_users}")
    print(f"Usuarios con al menos un churn: {users_with_churn}")
    print(f"Usuarios con múltiples churns: {users_with_multiple_churns} ({users_with_multiple_churns/total_users:.1%} del total)")

    # Mostrar distribución de número de churns
    churn_distribution = churn_counts.value_counts().sort_index()
    print("\nDistribución de número de churns por usuario:")
    for num_churns, count in churn_distribution.items():
        print(f"  {num_churns} churn(s): {count} usuarios ({count/total_users:.1%})")

    # Obtener los 5 usuarios con más churns para análisis detallado
    top_churners = multi_churn_users.head(5).index.tolist()
    print(f"\nTop 5 usuarios con más churns: {top_churners}")

    # Mostrar ejemplos detallados de los primeros 3 usuarios con múltiples churns
    print("\nEjemplos detallados de usuarios con múltiples churns:")
    examples = {}

    for i, user_id in enumerate(top_churners[:3]):
        user_data = df_target[df_target['advertiser_zrive_id'] == user_id].sort_values('period_int')
        churn_months = user_data[user_data['churn'] == 1]['period_int'].tolist()
        
        print(f"\nUsuario ID {user_id} - {len(churn_months)} churns en los meses: {churn_months}")
        print(user_data[['period_int', 'month_period', 'monthly_published_ads', 'monthly_contracted_ads', 'has_active_contract', 'churn']])
        
        examples[user_id] = {
            'data': user_data,
            'churn_months': churn_months
        }

    # Devolver información para uso posterior
    return {
        'multi_churn_users': multi_churn_users,
        'top_churners': top_churners,
        'examples': examples,
        'churn_distribution': churn_distribution
    }

In [62]:
find_users_with_multiple_churns(df_target)

Total de usuarios únicos: 5971
Usuarios con al menos un churn: 3862
Usuarios con múltiples churns: 1314 (22.0% del total)

Distribución de número de churns por usuario:
  1 churn(s): 2548 usuarios (42.7%)
  2 churn(s): 760 usuarios (12.7%)
  3 churn(s): 299 usuarios (5.0%)
  4 churn(s): 137 usuarios (2.3%)
  5 churn(s): 56 usuarios (0.9%)
  6 churn(s): 26 usuarios (0.4%)
  7 churn(s): 18 usuarios (0.3%)
  8 churn(s): 8 usuarios (0.1%)
  9 churn(s): 6 usuarios (0.1%)
  10 churn(s): 1 usuarios (0.0%)
  11 churn(s): 1 usuarios (0.0%)
  14 churn(s): 1 usuarios (0.0%)
  16 churn(s): 1 usuarios (0.0%)

Top 5 usuarios con más churns: [2888, 1795, 1760, 4689, 1890]

Ejemplos detallados de usuarios con múltiples churns:

Usuario ID 2888 - 16 churns en los meses: [202302, 202303, 202304, 202306, 202308, 202309, 202312, 202401, 202402, 202403, 202404, 202406, 202408, 202409, 202410, 202411]
       period_int month_period  monthly_published_ads  monthly_contracted_ads  \
2104       202301      202

{'multi_churn_users': advertiser_zrive_id
 2888    16
 1795    14
 1760    11
 4689    10
 1890     9
         ..
 2693     2
 2699     2
 2701     2
 2702     2
 6350     2
 Length: 1314, dtype: int64,
 'top_churners': [2888, 1795, 1760, 4689, 1890],
 'examples': {2888: {'data':        advertiser_zrive_id  period_int month_period  monthly_published_ads  \
   2104                  2888      202301      2023-01                     20   
   5541                  2888      202302      2023-02                     20   
   8928                  2888      202303      2023-03                     20   
   12329                 2888      202304      2023-04                     20   
   15757                 2888      202305      2023-05                     20   
   19132                 2888      202306      2023-06                     20   
   22520                 2888      202307      2023-07                     20   
   25857                 2888      202308      2023-08                    

In [63]:
df_target[df_target['advertiser_zrive_id'] == 2888]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
2104                  2888      202301      2023-01                     20   
5541                  2888      202302      2023-02                     20   
8928                  2888      202303      2023-03                     20   
12329                 2888      202304      2023-04                     20   
15757                 2888      202305      2023-05                     20   
19132                 2888      202306      2023-06                     20   
22520                 2888      202307      2023-07                     20   
25857                 2888      202308      2023-08                     20   
29067                 2888      202309      2023-09                     20   
32258                 2888      202310      2023-10                     20   
35441                 2888      202311      2023-11                     35   
38658                 2888      202312      2023-12                     35   
38659                 2888      202312      2023-12                     35   
41867                 2888      202401      2024-01                     46   
45081                 2888      202402      2024-02                     45   
48332                 2888      202403      2024-03                     40   
51662                 2888      202404      2024-04                     40   
51663                 2888      202404      2024-04                     40   
55063                 2888      202405      2024-05                     33   
58520                 2888      202406      2024-06                     35   
62015                 2888      202407      2024-07                     35   
65390                 2888      202408      2024-08                     35   
68749                 2888      202409      2024-09                     39   
72116                 2888      202410      2024-10                     37   
75428                 2888      202411      2024-11                     33   
78756                 2888      202412      2024-12                     31   

       monthly_contracted_ads  has_active_contract  churn  
2104                       20                 True    0.0  
5541                       20                 True    1.0  
8928                       20                 True    1.0  
12329                      20                 True    1.0  
15757                      20                 True    0.0  
19132                      20                 True    1.0  
22520                      20                 True    0.0  
25857                      20                 True    1.0  
29067                      20                 True    1.0  
32258                      20                 True    0.0  
35441                      35                 True    0.0  
38658                      35                 True    0.0  
38659                      35                 True    1.0  
41867                      50                 True    1.0  
45081                      50                 True    1.0  
48332                      50                 True    1.0  
51662                      50                 True    1.0  
51663                      50                 True    0.0  
55063                      35                 True    0.0  
58520                      35                 True    1.0  
62015                      35                 True    0.0  
65390                      35                 True    1.0  
68749                      35                 True    1.0  
72116                      35                 True    1.0  
75428                      35                 True    1.0  
78756                      35                 True    0.0

In [64]:
df_target[df_target['advertiser_zrive_id'] == 1890]

advertiser_zrive_id  period_int month_period  monthly_published_ads  \
34824                 1890      202311      2023-11                      2   
38044                 1890      202312      2023-12                      8   
38045                 1890      202312      2023-12                      8   
41255                 1890      202401      2024-01                      5   
44472                 1890      202402      2024-02                      6   
47730                 1890      202403      2024-03                     10   
51062                 1890      202404      2024-04                      6   
54455                 1890      202405      2024-05                      2   
57930                 1890      202406      2024-06                      9   
61438                 1890      202407      2024-07                     10   
64817                 1890      202408      2024-08                     10   
68179                 1890      202409      2024-09                      8   
71553                 1890      202410      2024-10                      9   

       monthly_contracted_ads  has_active_contract  churn  
34824                      10                 True    0.0  
38044                      10                 True    1.0  
38045                      10                 True    1.0  
41255                      10                 True    0.0  
44472                      10                 True    1.0  
47730                      10                 True    1.0  
51062                      10                 True    1.0  
54455                      10                 True    1.0  
57930                      10                 True    1.0  
61438                      10                 True    0.0  
64817                      10                 True    1.0  
68179                      10                 True    1.0  
71553                      10                 True    0.0